### PR TITLE
Adding AugMix implementation

### PR DESCRIPTION
**Pull request by datumbox**
Adding the [AugMix](https://arxiv.org/abs/1912.02781) data augmentation method. Inspired from the work on the [official repo](https://github.com/google-research/augmix/blob/9b9824c7c19bf7e72df2d085d97b99b3bfb00ba4/imagenet.py#L209).

### CODE REVIEW

File: torchvision/transforms/autoaugment.py [hunk 1]
```diff
@@ -458,3 +458,140 @@ def __repr__(self) -> str:
...
+            img = orig_img
+            if isinstance(fill, (int, float)):
+                fill = [float(fill)] * F.get_image_num_channels(img)
+            elif fill is not None:
+                fill = [float(f) for f in fill]
+        else:
+            img = self._pil_to_tensor(orig_img)
+
+        mixing_weights = torch._sample_dirichlet(torch.tensor([self.alpha] * self.mixture_width, device=img.device))
+        m = torch._sample_dirichlet(torch.tensor([self.alpha, self.alpha], device=img.device))[0]
```
Comment (datumbox) [inline torchvision/transforms/autoaugment.py:556]: We use the same `m` value for the entire batch.

File: torchvision/transforms/autoaugment.py [hunk 2]
```diff
@@ -458,3 +458,140 @@ def __repr__(self) -> str:
...
+        op_meta = self._augmentation_space(self._PARAMETER_MAX, F.get_image_size(img))
+
+        orig_dims = list(img.shape)
+        batch = img.view([1] * max(4 - img.ndim, 0) + orig_dims)
+        mix = torch.zeros_like(batch, dtype=torch.float)
+        for i in range(self.mixture_width):
+            aug = batch.clone()
+            depth = self.chain_depth if self.chain_depth > 0 else int(torch.randint(low=1, high=4, size=(1,)).item())
```
Comment (datumbox) [inline torchvision/transforms/autoaugment.py:564]: If negative (-1) we randomly sample a depth between `[1,3]`

**Comment by datumbox** [comment]
@hendrycks Thanks for the input! 

In the current implementation, we support it and make it optional to use the extra 4 transforms by setting `all_ops=True` on the constructor. Check this https://github.com/pytorch/vision/pull/5411#discussion_r804798735

Let me know if you have any other feedback on the implementation. The main changes are on the `autoaugment.py` file of this PR.

**Comment by datumbox** [comment]
@normster @hendrycks We intend to merge soon. If you have any thoughts on the comments above let us know, thanks!

File: torchvision/transforms/autoaugment.py [hunk 3]
```diff
@@ -458,3 +458,140 @@ def __repr__(self) -> str:
...
+            image. If given a number, the value is used for all bands respectively.
+    """
+
+    def __init__(
+        self,
+        severity: int = 1,
+        mixture_width: int = 3,
+        chain_depth: int = -1,
+        alpha: float = 1.0,
+        all_ops: bool = True,
```
[the remark this answers is not in the document]
Comment (normster) [reply torchvision/transforms/autoaugment.py:490]: These look good

File: torchvision/transforms/autoaugment.py [hunk 4]
```diff
@@ -458,3 +458,140 @@ def __repr__(self) -> str:
...
+        if isinstance(orig_img, Tensor):
+            img = orig_img
+            if isinstance(fill, (int, float)):
+                fill = [float(fill)] * F.get_image_num_channels(img)
+            elif fill is not None:
+                fill = [float(f) for f in fill]
+        else:
+            img = self._pil_to_tensor(orig_img)
+
+        mixing_weights = torch._sample_dirichlet(torch.tensor([self.alpha] * self.mixture_width, device=img.device))
```
[the remark this answers is not in the document]
Comment (normster) [reply torchvision/transforms/autoaugment.py:555]: We sample different mixing/m values per image in the original implementation. Since this is a torchvision transform, isn't it usually applied per image?

**Review by normster — COMMENTED** [review]
This looks great! I had a small question about per-image/per-batch processing that I noted inline but what you have should work fine in practice.

**Comment by datumbox** [comment]
@normster Thanks a lot for the feedback. In relation to the sampling of weights, I will benchmark your proposal. Out of curiosity, did you do any experiments on the sampling per image vs batch?

File: torchvision/transforms/autoaugment.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (datumbox) [reply torchvision/transforms/autoaugment.py:555]: Thanks for checking. In this implementation it depends on whether the user calls it by providing an image vs batch. We can certainly force sampling per image in all cases, but it's going to be slightly slower. I wonder what the accuracy implications of this choice is. If you have done experiments on this please let me know.

File: torchvision/transforms/autoaugment.py [hunk 5]
```diff
@@ -458,3 +458,150 @@ def __repr__(self) -> str:
...
+        """
+        fill = self.fill
+        if isinstance(orig_img, Tensor):
+            img = orig_img
+            if isinstance(fill, (int, float)):
+                fill = [float(fill)] * F.get_image_num_channels(img)
+            elif fill is not None:
+                fill = [float(f) for f in fill]
+        else:
+            img = self._pil_to_tensor(orig_img)
```
Comment (vfdev-5) [inline torchvision/transforms/autoaugment.py:557]: Later we may want to refactor this part of code that could be applicable to all other aug strategies...

Comment (datumbox) [reply torchvision/transforms/autoaugment.py:557]: Certainly. We would need to consider how Videos can be handled as well here too.

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @datumbox !
